### PR TITLE
redesign kvm qmp monitor

### PR DESCRIPTION
**Pull request by codefritzel**
This commit has changed the substructure of QMP communication.  The code has been modernised with typehints codestyle and better logging.

The inheritance hierarchy has been adapted so that communication with the QGA agent is now possible.

Furthermore, the QmpConnection class can now wait for certain events (see QmpConnection#wait_for_qmp_event).  This can also solve the issue #1749.

To minimise code changes, the old method 'Execute' refers to the new 'execute_qmp'.

### CODE REVIEW

File: lib/hypervisor/hv_kvm/monitor.py [hunk 1]
```diff
@@ -811,6 +771,21 @@ def GetCpuInformation(self):
 
     return self.Execute("query-cpus-fast")
 
+  @_ensure_connection
+  def GetHotpluggableCPUs(self):
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:775]: not sure why this method gets added here

File: lib/hypervisor/hv_kvm/monitor.py [hunk 2]
```diff
@@ -811,6 +771,21 @@ def GetCpuInformation(self):
...
+    """ Get CPU type which could be plugged
+        uses the query-hotpluggable-cpus which does not interrupt the guest
+    """
+
+    return self.Execute("query-hotpluggable-cpus")
+
+  @_ensure_connection
+  def QueryMemorySummary(self):
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:783]: same as above, why is this method added here?

File: lib/hypervisor/hv_kvm/monitor.py [hunk 3]
```diff
@@ -273,6 +366,30 @@ def __enter__(self):
...
+    self.send_qmp(message)
+
+    ret = self.get_qmp_response(command)
+    if command not in [self._QUERY_COMMANDS, self._CAPABILITIES_COMMAND]:
+      logging.debug("QMP Response: %s %s: %s\n", command, arguments, ret)
+    return ret
+
+  # as workaround for existing code
+  # this should be changed in the future!
+  def Execute(self, command, arguments=None):
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:390]: I think we should update all calls to `Execute()` to use `execute_qmp()` instead as they are only limited to the hv_kvm implementation and quite easy to spot/replace. I do not see much sense in adding a temporary alias for backwards compatibility.

File: lib/hypervisor/hv_kvm/monitor.py [hunk 4]
```diff
@@ -591,25 +568,11 @@ def _GetPCIDevices(self):
...
     """
-    self._check_connection()
     pci = self.Execute("query-pci")
     bus = pci[0]
     devices = bus["devices"]
     return devices
 
-
-  @_ensure_connection
-  def ResizeBlockDevice(self, disk_id: str, new_size: int):
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:602]: Why is this method deleted?

File: lib/hypervisor/hv_kvm/monitor.py [hunk 5]
```diff
@@ -664,7 +626,7 @@ def GetFreePCISlot(self):
 
     """
     slots = bitarray(constants.QEMU_PCI_SLOTS)
-    slots.setall(False) # pylint: disable=E1101
+    slots.setall(False)  # pylint: disable=E1101
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:629]: extra whitespace

File: lib/hypervisor/hv_kvm/monitor.py [hunk 6]
```diff
@@ -732,7 +692,7 @@ def HasDynamicAutoReadOnly(self):
       return True
 
     return any([x for x in schema
-                if "dynamic-auto-read-only" in x.get("features",[])])
+                if "dynamic-auto-read-only" in x.get("features", [])])
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:695]: extra whitespace

File: lib/hypervisor/hv_kvm/monitor.py [hunk 7]
```diff
@@ -917,3 +889,46 @@ def _RemoveFdset(self, fdset):
       # succeeded, the whole hot-add action will fail and the runtime file will
       # not be updated which will make the instance non migrate-able
       logging.info("Removing fdset with id %s failed: %s", fdset, err)
+
+
+class QgaConnection(QemuMonitorSocket):
```
Comment (rbott) [inline lib/hypervisor/hv_kvm/monitor.py:894]: Not entirely sure why this is here, AFAIK there would be nothing using this code right now, correct? Maybe this should be added as a separate commit.

**Review by rbott — COMMENTED** [review]
Hi @codefritzel, thanks for taking on this important task :-) A rework of the QMP code has been long overdue (especially since it has become more important in the past years).

I only have some preliminary comments for now, I need some more time to look into this more thoroughly.

I have started a QA run on [stargazer.ganeti.org](https://stargazer.ganeti.org/), logs/details will be available [here](https://stargazer.ganeti.org/2f42c2948908e40a5c9c86ca8bd0eb4d88c8cacd/) once it has finished.
*Edit*: it successfully finished 😃 

Regarding typehints: I am a huge fan of Python typehints. However, I am not sure if this would be the right way to introduce them to Ganeti.

I would rather suggest to add a section to https://github.com/ganeti/ganeti/blob/master/CONTRIBUTING.md with Python coding/style guidance, reflecting the current state of affairs (in a separate PR). From there, we can agree on changes (e.g. typehints, line length, method/function name conventions etc.) and also how to carry them out throughout the project. What would be your views on this @saschalucas and @apoikos?

File: lib/hypervisor/hv_kvm/monitor.py (hunk 5 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:629]: pycharm has reported this

File: lib/hypervisor/hv_kvm/monitor.py (hunk 6 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:695]: pycharm has reported this

File: lib/hypervisor/hv_kvm/monitor.py (hunk 1 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py]: removed

File: lib/hypervisor/hv_kvm/monitor.py (hunk 2 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py]: removed

File: lib/hypervisor/hv_kvm/monitor.py (hunk 7 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:890]: Yes, this is not currently used. In the future, this will be used to communicate with the guest agent via qmp.

**Comment by codefritzel** [comment]
Hi @rbott,
some things have sneaked in from a parallel new feature development. 

> Regarding typehints: I am a huge fan of Python typehints. However, I am not sure if this would be the right way to introduce them to Ganeti.

Yes, it is definitely the right way, because at some point the code has to be renewed to prevent it from dying out.
We have to start somewhere. Why not at this one?

>I would rather suggest to add a section to https://github.com/ganeti/ganeti/blob/master/CONTRIBUTING.md with Python coding/style guidance, reflecting the current state of affairs (in a separate PR). From there, we can agree on changes (e.g. typehints, line length, method/function name conventions etc.) and also how to carry them out throughout the project. What would be your views on this @saschalucas and @apoikos?

I am of the opinion that new and especially large code changes should be made in modern Python code (e.g. typehints, style, protocols, data classes, ...)

File: lib/hypervisor/hv_kvm/monitor.py (hunk 3 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:390]: I would leave this as it is for the time being, as I'm still working on another feature that needs to be added first.
Then i would rename this and all other methods according to the desired style

**Comment by saschalucas** [comment]
> Regarding typehints: I am a huge fan of Python typehints. However, I am not sure if this would be the right way to introduce them to Ganeti.
> 
> I would rather suggest to add a section to https://github.com/ganeti/ganeti/blob/master/CONTRIBUTING.md with Python coding/style guidance, reflecting the current state of affairs (in a separate PR). From there, we can agree on changes (e.g. typehints, line length, method/function name conventions etc.) and also how to carry them out throughout the project. What would be your views on this @saschalucas and @apoikos?

Thanks @rbott for taking review. I agree that some coding and style standards should be developed and published in `CONTRIBUTING.md`. But I think it does not hurt to have non standardized/agreed hints in this PR.

File: lib/hypervisor/hv_kvm/monitor.py (hunk 5 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:640]: PEP 8: E261 at least two spaces before inline comment
https://peps.python.org/pep-0008/#inline-comments

File: lib/hypervisor/hv_kvm/monitor.py (hunk 6 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:706]: PEP 8: E231 missing whitespace after ','
https://www.flake8rules.com/rules/E231.html

File: lib/hypervisor/hv_kvm/monitor.py (hunk 5 above)
Comment (saschalucas) [reply lib/hypervisor/hv_kvm/monitor.py:635]: Acceptable argument.

File: lib/hypervisor/hv_kvm/monitor.py (hunk 6 above)
Comment (saschalucas) [reply lib/hypervisor/hv_kvm/monitor.py]: Also acceptable argument. I should definitely create an issue to not forget writing style guidelines ...

File: lib/hypervisor/hv_kvm/__init__.py [hunk 8]
```diff
@@ -69,8 +69,8 @@
 from ganeti.hypervisor import hv_base
 from ganeti.utils import wrapper as utils_wrapper
 
-from ganeti.hypervisor.hv_kvm.monitor import QmpConnection, QmpMessage, \
-                                             MonitorSocket
+from ganeti.hypervisor.hv_kvm.monitor import QmpConnection, QmpMessage \
+
```
Comment (saschalucas) [inline lib/hypervisor/hv_kvm/__init__.py:73]: The line continuation and extra newline here looks unnecessary?

File: lib/hypervisor/hv_kvm/monitor.py [hunk 9]
```diff
@@ -632,10 +590,20 @@ def _GetBlockDevices(self):
     @return: Info about the virtual disks of the instance.
 
     """
-    self._check_connection()
-    devices = self.Execute("query-block")
+    devices = self.execute_qmp("query-block")
     return devices
 
+  @_ensure_connection
+  def ResizeBlockDevice(self, disk_id: str, new_size: int):
```
Comment (saschalucas) [inline lib/hypervisor/hv_kvm/monitor.py:597]: hasn't this been brought by PR #1731 / commit ceff246e30? ATM it puzzles me why this is in the diff of this PR...

File: lib/hypervisor/hv_kvm/monitor.py [hunk 10]
```diff
@@ -61,18 +61,15 @@ class QmpCommandNotSupported(errors.HypervisorError):
...
+
+class QmpMessage:
   """QEMU Messaging Protocol (QMP) message.
 
   """
-  def __init__(self, data):
-    """Creates a new QMP message based on the passed data.
-
-    """
-    if not isinstance(data, dict):
```
[the remark this answers is not in the document]
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:73]: In my opinion, the check is unnecessary because:
1. the typehint tell you what is expected
2. such checks are usually useful for user input, which is not the case here. Only the programmer can pass the wrong type.
3. if it makes sense at this point, why hasn't it been done in the many other places?
4. even if performance is not an argument in Python, but if is a jump command that should be avoided if possible

If such a check should remain, then an assert would make sense, which is only active in debug mode.

File: lib/hypervisor/hv_kvm/monitor.py (hunk 9 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py:597]: When solving a merge conflict, this part was removed... This was added again in 4fc8896dfe73a3bc9dd65973dd56dd42ce0d246c

Comment (saschalucas) [reply lib/hypervisor/hv_kvm/monitor.py:597]: I see, it was accidentally removed and the reinserted again, but a view lines below. Originally it was before `_HasPCIDevice()`, now it is after ` _GetBlockDevices()`. Solved as reordering.

File: lib/hypervisor/hv_kvm/monitor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (saschalucas) [reply lib/hypervisor/hv_kvm/monitor.py:73]: ACK

File: lib/hypervisor/hv_kvm/monitor.py (hunk 7 above)
Comment (saschalucas) [reply lib/hypervisor/hv_kvm/monitor.py:888]: I'm fine with leaving the `QgaConnection()` class here, it still serves the purpose of the PRs description "communication with the QGA agent is now possible". Maybe next time split PRs more granular.

File: lib/hypervisor/hv_kvm/monitor.py (hunk 9 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py]: fixed

**Review by saschalucas — APPROVED** [review]
Thanks @codefritzel. According to @rbott QA suite was successful. Other issues were resolved at a level that we can achieve.

Please merge your changes (squash).

File: lib/hypervisor/hv_kvm/monitor.py (hunk 7 above)
Comment (rbott) [reply lib/hypervisor/hv_kvm/monitor.py:888]: I still think this shouldn't be here or at least there should be a design document that explains what we want to achieve with agent support.

**Comment by rbott** [comment]
I will start a new QA run to cover the latest changes, please wait for that to finish.

**Comment by codefritzel** [comment]
> I will start a new QA run to cover the latest changes, please wait for that to finish.

Hi @rbott,
i squashed the commits into one. Can you please run the qa now?

**Comment by rbott** [comment]
Sorry for the delay, the QA suite is running: https://stargazer.ganeti.org/db0965786823503326364b4642ff6fb43ffa3468/

It will take ~5 to 6 hours, unless it fails.You can see the status in the [overview](https://stargazer.ganeti.org/).

File: lib/hypervisor/hv_kvm/monitor.py (hunk 7 above)
Comment (codefritzel) [reply lib/hypervisor/hv_kvm/monitor.py]: I moved the QgaConnection class in another branch, so this is out for now.

**Comment by codefritzel** [comment]
@saschalucas @rbott 
I have removed the qga class for the time being. I don't think a new qa run is necessary.
Can one of you please merge it?

**Comment by saschalucas** [comment]
> I have removed the qga class for the time being.

good.

> I don't think a new qa run is necessary.

Agreed.

> Can one of you please merge it?

Since you have write access now, it should be possible to merge your PRs your self :-). The protecting step is the approval not the merge itself.

**Comment by saschalucas** [comment]
> Sorry for the delay, the QA suite is running: https://stargazer.ganeti.org/db0965786823503326364b4642ff6fb43ffa3468/
> 
> It will take ~5 to 6 hours, unless it fails.You can see the status in the [overview](https://stargazer.ganeti.org/).

QA was successful here:
```text
2024-08-16 17:18:53 UTC 	finished 	positive-herman 	kvm-drbd_file_sharedfile-bridged 	Debian/Bookworm 	codefritzel/ganeti:redesign_qmp 	05:49:57
```

**Comment by thielema** [comment]
On Wed, 21 Aug 2024, codefritzel wrote:

> @saschalucas @rbott
> I have removed the qga class for the time being. I don't think a new qa run is necessary.
> Can one of you please merge it?

Btw. is this something that interacts with the KVM UserShutdown issue?

https://github.com/ganeti/ganeti/issues/1760

**Review by rbott — APPROVED** [review]

**Comment by codefritzel** [comment]
> On Wed, 21 Aug 2024, codefritzel wrote: @saschalucas @rbott I have removed the qga class for the time being. I don't think a new qa run is necessary. Can one of you please merge it?
> Btw. is this something that interacts with the KVM UserShutdown issue? #1760

Hi @thielema,
I think the qmp part of the kvmd is implemented in Haskell.
